### PR TITLE
Use `atomicAdd` for `bfloat16` in Ampere and above

### PR DESCRIPTION
**Pull request by eqy**
WIP to fix extremely slow `scatter_add` issue vs. fp16. The current changes seem to improve performance, but it still appears to lag behind the fp16 equivalent.

CC @ngimel @ptrblck

### CODE REVIEW

File: aten/src/ATen/cuda/Atomic.cuh [hunk 1]
```diff
@@ -6,6 +6,10 @@
 
 #include <ATen/NumericUtils.h>
 
+#if !(defined(USE_ROCM) || ((defined(CUDA_VERSION) && CUDA_VERSION < 11000) || (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < 800))))
+#include <cuda_bf16.h>
```
Comment (eqy) [inline aten/src/ATen/cuda/Atomic.cuh:10]: I'm not sure why this is needed, if it it removed, the build fails with the complaint that `__nv_bfloat16` isn't defined when the included `c10/util/BFloat16.h` header should also include it...

File: aten/src/ATen/cuda/Atomic.cuh [hunk 2]
```diff
@@ -222,10 +226,15 @@ static inline  __device__ at::Half gpuAtomicAdd(at::Half *address, at::Half val)
...
-                                    [](at::BFloat16 bsum, at::BFloat16 val) {
-                                      return bsum + val;
-                                    });
+#if defined(USE_ROCM) || ((defined(CUDA_VERSION) && CUDA_VERSION < 11000) || (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < 800)))
+return AtomicFPOp<at::BFloat16>()(address, val,
+                                  [](at::BFloat16 bsum, at::BFloat16 val) {
+                                    return bsum + val;
+                                  });
+#else
+  __nv_bfloat16 r = atomicAdd(reinterpret_cast<__nv_bfloat16*>(address), *reinterpret_cast<__nv_bfloat16*>(&val));
```
Comment (eqy) [inline aten/src/ATen/cuda/Atomic.cuh:235]: This is very clunky and I should be able to use the functions in `c10/util/BFloat16-inl.h` but I was tripping over the syntax.

**Comment by ngimel** [comment]
What are the perf numbers you are getting with this?

**Comment by eqy** [comment]
Better than before but still very much the wrong order of magnitude according to the microbenchmark (on A6000):
before
```
INFO:root:matmul tf32: True
INFO:root:cudnn tf32: True
INFO:root:matmul fp16 reduction: True
INFO:root:Input_size 2048, dtype torch.float16
INFO:root:batch size 128, forward 0.072292057, backward 0.034451169
INFO:root:batch size 256, forward 0.00296005, backward 0.016321898
INFO:root:batch size 512, forward 0.005347602, backward 0.029631895
INFO:root:****
INFO:root:Input_size 2048, dtype torch.bfloat16
INFO:root:matmul tf32: True
INFO:root:cudnn tf32: True
INFO:root:matmul fp16 reduction: True
INFO:root:batch size 128, forward 0.049289932, backward 3.460327631
INFO:root:batch size 256, forward 0.002965743, backward 5.031641967
INFO:root:batch size 512, forward 0.005354016, backward 4.805066426
```
after
```INFO:root:matmul tf32: True
INFO:root:cudnn tf32: True
INFO:root:matmul fp16 reduction: True
INFO:root:Input_size 2048, dtype torch.float16
INFO:root:batch size 128, forward 0.071989492, backward 0.040035025
INFO:root:batch size 256, forward 0.002961816, backward 0.016255412
INFO:root:batch size 512, forward 0.005348356, backward 0.029601357
INFO:root:****
INFO:root:Input_size 2048, dtype torch.bfloat16
INFO:root:matmul tf32: True
INFO:root:cudnn tf32: True
INFO:root:matmul fp16 reduction: True
INFO:root:batch size 128, forward 0.047254223, backward 1.572099484
INFO:root:batch size 256, forward 0.002972221, backward 2.16478994
INFO:root:batch size 512, forward 0.00535267, backward 2.945632733
```

**Comment by ngimel** [comment]
would bfloat2 atomicAdd help? (This is what's used for half) https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH____BFLOAT162__ARITHMETIC.html#group__CUDA__MATH____BFLOAT162__ARITHMETIC_1g550f52c89d672213390e9bfd8a3c42bf
You can check your generated sass with bfloat atomicAdd, I suspect it still generates CAS

**Comment by eqy** [comment]
> would bfloat2 atomicAdd help? (This is what's used for half) https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH____BFLOAT162__ARITHMETIC.html#group__CUDA__MATH____BFLOAT162__ARITHMETIC_1g550f52c89d672213390e9bfd8a3c42bf You can check your generated sass with bfloat atomicAdd, I suspect it still generates CAS

Not sure what this entails, naively changing the cast types yields a misaligned address error,,,

Will check the generated code.

**Comment by ngimel** [comment]
Yeah you cannot naively change the type, you should use the approach similar to fastAtomicAdd for fp16, where you apply atomicAdd to the properly aligned reinterpret-casted half2 (or bfloat162) pointer, adding 0 to the second half. 
However, I've checked with godbolt, and CAS is used both in bfloat16 and bfloat162 atomic versions, so I don't have high hopes.

**Comment by alihassanijr** [comment]
I can confirm that I already tried reimplementing `fastSpecializedAtomicAdd` (see below) with `__nv_bfloat162` and it's still significantly slower in my use case.
https://github.com/pytorch/pytorch/blob/9d1155235b82a83b49b7645db5292491ea81dacf/aten/src/ATen/native/cuda/KernelUtils.cuh#L25-L64

One more interesting thing that I keep running into is the lack of a defined typecast between `c10::BFloat16` and `__nv_bfloat16`. `c10::Half` casts to `__half` without a hitch, but in the case of `bfloat16` I've had to just manually reinterpret cast to avoid the issue.

**Comment by alihassanijr** [comment]
I may be completely reading this wrong, but it seems like all atomic operations for bfloat16 are implemented with CAS, as opposed to float16 and float162 which seem to have a dedicated add instruction:
[https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-atom](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-atom)

I ended up switching the kernel I'm struggling with to float32 at all times (since it's just a backwards kernel for a really small weight tensor, so return a float32 tensor isn't too big of a deal). It is just a temporary solution for my use case to drop that 10ms latency back down to 100us. It's almost the same latency as float16 now, but this particular kernel doesn't even use half2 right now, so no harm done.

That said, I will try and look into this some more and see if I can figure something else out.

**Comment by alihassanijr** [comment]
I think CUDA v11.8 might help resolve the speed issue by the way.

Notice how CUDA v11.7 only had a CAS instruction for b16:
```
atom{.sem}{.scope}{.space}.cas.b16 d, [a], b, c;
```
Source: https://docs.nvidia.com/cuda/archive/11.7.1/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-atom

However, in v11.8 we get this nice little addition which is all I think is needed
```
atom{.sem}{.scope}{.space}.add.noftz{.level::cache_hint}.bf16    d, [a], b{, cache-policy};
atom{.sem}{.scope}{.space}.add.noftz{.level::cache_hint}.bf16x2  d, [a], b{, cache-policy};
```
Source: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-atom

**Comment by eqy** [comment]
I'm seeing the following PTX being generated with just a standalone call to `atomicAdd` on `sm90`:
`atom.add.noftz.f16 %rs2,[%rd1],%rs1;` (fp16)
`atom.add.noftz.bf16 %rs2,[%rd1],%rs1;` (bf16)

However, the slowdown remains so I'm suspecting that it might be another kernel/operation that is slow.

**Comment by alihassanijr** [comment]
@eqy Interesting. I'm assuming you're on 11.8? I thought 12 was going to be the first version to support `sm90`.

**Comment by eqy** [comment]
Yes, this is on 11.8 but I'm not 100% sure my build is correct and need to recheck.

**Comment by alihassanijr** [comment]
Yeah I just noticed there's a new NGC image with the latest torch. I'm going to give it a shot with my example and see if I observe anything different.

**Comment by alihassanijr** [comment]
So I'm still confirming, because my environment's a bit of a mess, but in my case the issue is resolved without any changes to torch (not using the vectroized bfloat2 ops or anything, just plain `fastAtomicAdd` from `aten::native`).
I was getting upwards of 10ms latency on a really simple kernel with an atomicAdd, and just under 1ms with `half` and `float`.
Now I'm getting that same <1ms with bfloat as well.

**Comment by eqy** [comment]
Thanks for checking on your end @alihassanijr . On `sm_90` after more testing I noticed that the `__half2` codepath was important for performance, and removing it regressed performance to be about the same as `__nv_bfloat16`. After adding the `__nv_bfloat162` equivalent I'm observing comparable performance in the benchmark from @ngimel :
```
INFO:root:matmul tf32: True
INFO:root:cudnn tf32: True
INFO:root:matmul fp16 reduction: True
INFO:root:Input_size 2048, dtype torch.float16
INFO:root:batch size 128, forward 0.186238571, backward 0.032990263
INFO:root:batch size 256, forward 0.002154732, backward 0.013528392
INFO:root:batch size 512, forward 0.003322089, backward 0.02547441
INFO:root:****                                                                                                                                                                                                                               INFO:root:Input_size 2048, dtype torch.bfloat16
INFO:root:matmul tf32: True
INFO:root:cudnn tf32: True                                                                                                                                                                                                                   INFO:root:matmul fp16 reduction: True
INFO:root:batch size 128, forward 0.006974035, backward 0.007897218
INFO:root:batch size 256, forward 0.002154933, backward 0.013497994
INFO:root:batch size 512, forward 0.003340705, backward 0.025482254
```

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**